src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,106 @@
+-- SPDX-FileCopyrightText: 2020 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+-- | Proposal related entrypoints
+module Lorentz.Contracts.BaseDAO.Proposal
+  ( propose
+  ) where
+
+import Lorentz
+import Lorentz.Contracts.BaseDAO.Types
+
+checkIsProposalValid
+  :: forall store pm. (IsoValue pm, StorageC store pm)
+  => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+checkIsProposalValid = do
+  duupX @2
+  stackType @[store, ProposeParams pm, store]
+  stToField #sRejectedProposalReturnValue
+  toNamed #sRejectedProposalReturnValue
+  stackType @["sRejectedProposalReturnValue" :! Natural, ProposeParams pm, store]
+
+  duupX @2
+  toFieldNamed #ppProposalPriceInToken
+  stackType @["ppProposalPriceInToken" :! Natural, "sRejectedProposalReturnValue" :! Natural, ProposeParams pm, store]
+
+  -- TODO: `sRejectedProposalReturnValue` should be lambda instead of static nat
+  if #ppProposalPriceInToken <. #sRejectedProposalReturnValue then
+    failCustom_ #fAIL_PROPOSAL_CHECK -- TODO: placeholder error
+  else nop
+
+checkSenderUnfrozenToken
+  :: forall store pm. (IsoValue pm, StorageC store pm)
+  => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+checkSenderUnfrozenToken = do
+  duupX @2
+  push (0 :: Natural) -- unfrozen token_id
```

Comment:
It'd be nice to have `unfrozenTokenId :: TokenId` defined somewhere. Same for `frozenTokenId`.

Reply:
I have added this in `BaseDAO.Types`.